### PR TITLE
Json parse improvements

### PR DESCRIPTION
**Pull request by andreaTP**

### CODE REVIEW

File: components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java [hunk 1]
```diff
@@ -348,15 +271,10 @@ private <T extends Parsable> void assignFieldValues(
...
-                    final Consumer<Parsable> onBefore = this.onBeforeAssignFieldValues;
-                    final Consumer<Parsable> onAfter = this.onAfterAssignFieldValues;
-                    fieldDeserializer.accept(
-                            new JsonParseNode(fieldValue) {
-                                {
-                                    this.setOnBeforeAssignFieldValues(onBefore);
-                                    this.setOnAfterAssignFieldValues(onAfter);
-                                }
-                            });
+                    JsonParseNode itemNode = new JsonParseNode(fieldValue);
```
Comment (baywet) [inline components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java:274]:
```suggestion
                    final JsonParseNode itemNode = new JsonParseNode(fieldValue);
```

File: components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java [hunk 2]
```diff
@@ -248,40 +198,13 @@ public T next() {
...
-                                }
-                            };
-                        }
-                    };
-
-            for (T elem : iterable) {
-                result.add(elem);
+            while (sourceIterator.hasNext()) {
+                JsonElement item = sourceIterator.next();
+                JsonParseNode itemNode = new JsonParseNode(item);
```
Comment (baywet) [inline components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java:204]:
```suggestion
                final JsonElement item = sourceIterator.next();
                final JsonParseNode itemNode = new JsonParseNode(item);
```

File: components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java [hunk 3]
```diff
@@ -203,38 +179,12 @@ public T next() {
...
```
Comment (baywet) [inline components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java:184]:
```suggestion
                final JsonElement item = sourceIterator.next();
                final JsonParseNode itemNode = new JsonParseNode(item);
```

File: components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java [hunk 4]
```diff
@@ -125,69 +158,12 @@ public JsonParseNode(@Nonnull final JsonElement node) {
...
```
Comment (baywet) [inline components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java:163]:
```suggestion
                final JsonElement item = sourceIterator.next();
                final JsonParseNode itemNode = new JsonParseNode(item);
```

File: components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java [hunk 5]
```diff
@@ -125,69 +158,12 @@ public JsonParseNode(@Nonnull final JsonElement node) {
...
-                                    }
-                                }
-                            };
-                        }
-                    };
-
-            for (T elem : iterable) {
-                result.add(elem);
+            while (sourceIterator.hasNext()) {
+                JsonElement item = sourceIterator.next();
```
Comment (baywet) [inline components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java:162]: we have those 4 first lines at 4 different places, do you think we could use a method to get the JsonParseNode from a JsonElement and set the the before/after assign field values callbacks?

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for cleaning this up, a couple of suggestions here.

File: components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java (hunk 1 above)
Comment (andreaTP) [reply components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java:274]: I accept those suggestions, but I appreciate that this is a style coming strongly from dotnet.
In java codebases, I have not seen much declaring `final` in function bodies, it doesn't hurt though.

File: components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java (hunk 5 above)
Comment (andreaTP) [reply components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java]: no strong opinion here, I don't know how much those routines are expected to change and, possibly, diverge over time.

Comment (baywet) [reply components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java]: I'd say not at all, besides the last line which is different for each block

Comment (andreaTP) [reply components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java]: ok, let me refactor this too

Comment (andreaTP) [reply components/serialization/json/src/main/java/com/microsoft/kiota/serialization/JsonParseNode.java]: refactored the iteration on the Array, I noticed only after that's slight different from what you asked for, but I think it reads pretty well now ... let me know

**Review by baywet — APPROVED** [review]
Thanks for making the changes!

**Pull request merged**